test/bigquery.ts

```diff
@@ -708,26 +735,207 @@ describe('BigQuery', () => {
+          query: QUERY_STRING,
+          params: NAMED_PARAMS,
+          types: INVALID_TYPES,
+        });
+      }, /Type not provided for parameter: key/);
+      done();
+    });
+
+    it('should throw for invalid type structure provided for named params', done => {
+      assert.throws(() => {
+        bq.createQueryJob({
+          query: QUERY_STRING,
+          params: NAMED_PARAMS,
+          types: POSITIONAL_PARAMS,
+        });
+      }, /Provided types must match the value type passed to `params`/);
+      done();
+    });
+
+    it('should throw for invalid provided types with positional params', done => {
+      const INVALID_TYPES = ['invalid'];
+
+      assert.throws(() => {
+        bq.createQueryJob({
+          query: QUERY_STRING,
+          params: POSITIONAL_PARAMS,
+          types: INVALID_TYPES,
+        });
+      }, /Invalid type provided./);
+      done();
+    });
+
+    it('should throw for invalid type structure provided for positional params', done => {
+      assert.throws(() => {
+        bq.createQueryJob({
+          query: QUERY_STRING,
+          params: POSITIONAL_PARAMS,
+          types: NAMED_TYPES,
+        });
+      }, /Provided types must match the value type passed to `params`/);
+      done();
+    });
+
+    it('should throw for incorrect number of types provided for positional params', done => {
+      const ADDITIONAL_TYPES = ['string', 'string'];
+      assert.throws(() => {
+        bq.createQueryJob({
+          query: QUERY_STRING,
+          params: POSITIONAL_PARAMS,
+          types: ADDITIONAL_TYPES,
+        });
+      }, /Incorrect number of parameter types provided./);
+      done();
+    });
+
+    it('should throw for empty array without provided types', done => {
+      bq.createJob = (reqOpts: JobOptions) => {
+        done();
+      };
+
+      assert.throws(() => {
+        bq.createQueryJob({
+          query: QUERY_STRING,
+          params: [],
+        });
+      }, /Type must be provided for empty array./);
+      done();
+    });
+
+    it('should throw for null value without provided types', done => {
+      bq.createJob = (reqOpts: JobOptions) => {
+        done();
+      };
+
+      assert.throws(() => {
+        bq.createQueryJob({
+          query: QUERY_STRING,
+          params: null,
+        });
+      }, /Type must be provided for null values./);
+      done();
+    });
+
+    it('should return correct types', () => {});
   });
 
   describe('valueToQueryParameter_', () => {
     it('should get the type', done => {
       const value = {};
 
-      sandbox.stub(BigQuery, 'getType_').callsFake(value_ => {
-        assert.strictEqual(value_, value);
-        setImmediate(done);
-        return {
-          type: '',
-        };
-      });
+      sandbox
+        .stub(BigQuery, 'getTypeDescriptorFromValue_')
+        .callsFake(value_ => {
+          assert.strictEqual(value_, value);
+          setImmediate(done);
+          return {
```

Comment:
We need to follow this through to verify this object we return is used as we expect.